### ISSUE TITLE
Error: "WITH-EDITOR:: -c: line 0: unexpected EOF while looking for matching `''

### ISSUE THREAD

**Issue by svend** [magit/with-editor#40]
I'm getting the following error in the `*Async Shell Command*` buffer with version ee39f232f2daf997498a4030d104c35ae397f457 of with-editor when runing `(with-editor-async-shell-command "pass edit foobar") `. (pass command is from https://www.passwordstore.org/).

```
"WITH-EDITOR:: -c: line 0: unexpected EOF while looking for matching `''
"WITH-EDITOR:: -c: line 1: syntax error: unexpected end of file
Password unchanged.
```

The following also appears in the `*Messages*` buffer.

```
#<process Shell>
EDITOR=sh\ -c\ \'echo\ \"WITH-EDITOR\:\ \$\$\ OPEN\ \$0\"\;\ sleep\ 604800\ \&\ sleep\=\$\!\;\ trap\ \"kill\ \$sleep\;\ exit\ 0\"\ USR1\;\ trap\ \"kill\ \$sleep\;\ exit\ 1\"\ USR2\;\ wait\ \$sleep\' pass edit foobar: exited abnormally with code 1.
```

If I checkout the commit before 99d3278b1c79718de16dd4f57dcc8c4aa31a4051, the same command works.

I am running GNU Emacs 26.0.60 on macOS.

**Comment by tarsius** [comment]
Just in case, try with `crontab -e` instead of `pass edit foobar`.

**Comment by svend** [comment]
`crontab -e` works. I'll try to figure out what is different with `pass`.

**Comment by myrjola** [comment]
I have the same issue. I didn't figure out the root cause. The with-editor package has worked before with password-store.el. However, now there may be some kind of quotation issue. As a workaround I downgraded password-store.el to the commit before the with-editor support. Specifically, I reverted to commit [38ec1c72e29c872ec0cdde82f75490640d4019bf](https://git.zx2c4.com/password-store/commit/contrib/emacs/password-store.el?id=38ec1c72e29c872ec0cdde82f75490640d4019bf).

**Comment by DamienCassou** [comment]
I have exactly the same problem.

**Comment by tarsius** [comment]
I believe that `with-editor` isn't the cause here but that various tools that use `$EDITOR` fail if they encounter quotes in the value of that variable. `systemctl` seems to be [another](https://github.com/magit/with-editor/issues/31#issuecomment-327283435).

**Comment by DamienCassou** [comment]
The issue arrived with with-editor commit 99d3278b1c79718de16dd4f57dcc8c4aa31a4051. You can reproduce with this script in /tmp/pass

```bash
#!/usr/bin/env bash

${EDITOR} /home/cassou/.bashrc
```

and this form: `(with-editor-async-shell-command "/tmp/pass")`.

**Comment by tarsius** [comment]
> You can reproduce with this script

You have to use:

```bash
#!/usr/bin/env bash

eval "${EDITOR}" /home/cassou/.bashrc
```

> The issue arrived with with-editor commit 99d3278b1c79718de16dd4f57dcc8c4aa31a4051.

This commit didn't break anything by itself but because it removes a code-path that was previously used in most cases it causes another code-path to be used - and that exhibits this issue. But as I said, I believe the issue is that like your script many tools do not properly deal with spaces and/or quotes in `$EDITOR`.

**Comment by DamienCassou** [comment]
Is there any workaround we could have in `with-editor` or do we have to patch all shell scripts we want to use with `with-editor`?

**Comment by tarsius** [comment]
I could revert 99d3278b1c79718de16dd4f57dcc8c4aa31a4051 or add an option so that users could choose which defect bothers them less. It would be best though if `emacsclient` were modified to pass along the environment in all cases.

**Comment by svend** [comment]
I've submitted a patch for password-store to use eval.

**Comment by tarsius** [comment]
I have reverted the fix for #23, thus fixing the issue brought up here. I have also added an option for users who don't care about these issues, but don't like #23.
